### PR TITLE
Settings preprocessor for runtime

### PR DESCRIPTION
**Pull request by lasote**
- Should be opt-in or opt-out maybe?
- Related with #1497

### CODE REVIEW

File: conans/client/loader.py [hunk 1]
```diff
@@ -70,8 +71,11 @@ def __init__(self, runner, settings, profile):
         @param cached_env_values: EnvValues object
         """
         self._runner = runner
+
+        assert isinstance(settings, Settings)
         settings.values = profile.settings_values
-        assert settings is None or isinstance(settings, Settings)
+        preprocess(settings, output)
```
Comment (memsharded) [inline conans/client/loader.py]: Why this responsability inside the "Loader"? I think it should be independent, and the settings provided to the loader should be already complete and defined.

File: conans/client/manager.py [hunk 2]
```diff
@@ -185,7 +185,8 @@ def _get_graph_builder(self, loader, update, remote_proxy):
         return graph_builder
 
     def _get_deps_graph(self, reference, profile, filename, current_path, remote_proxy):
-        loader = ConanFileLoader(self._runner, self._client_cache.settings, self._profile_with_defaults(profile))
+        loader = ConanFileLoader(self._runner, self._client_cache.settings,
+                                 self._profile_with_defaults(profile), self._user_io.out)
```
Comment (memsharded) [inline conans/client/manager.py]: Maybe the right place could be the ``_profile_with_defaults()``

File: conans/client/settings_preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+def fill_runtime(settings, output):
+    try:
+        if settings.compiler == "Visual Studio":
+            if settings.get_safe("compiler.runtime") is None:
+                settings.compiler.runtime = "MDd" if settings.get_safe("build_type") == "Debug" \
+                                                  else "MD"
+                output.info("Setting 'compiler.runtime' not declared, automatically "
+                            "adjusted to '%s'" % settings.compiler.runtime)
+    except InvalidConanSettingField:  # If the settings structure doesn't match these general
```
Comment (memsharded) [inline conans/client/settings_preprocessor.py]: The usage of this Exception type seems a bit weak to me. I would have used just safe checks of settings, no Exception capture: Just an opinion, not really sure about best approach.

File: conans/client/settings_preprocessor.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+def preprocess(settings, output):
+    fill_runtime(settings, output)
+
+
+def fill_runtime(settings, output):
+    try:
+        if settings.compiler == "Visual Studio":
+            if settings.get_safe("compiler.runtime") is None:
+                settings.compiler.runtime = "MDd" if settings.get_safe("build_type") == "Debug" \
```
Comment (memsharded) [inline conans/client/settings_preprocessor.py]: Maybe also add the check for MT/MTd (raise if not matches build_type)?

File: conans/client/settings_preprocessor.py (hunk 3 above)
Comment (memsharded) [inline conans/client/settings_preprocessor.py]: I guess it is somewhat related to your question about opt-it, opt-out. We need to make sure this doesn't block users, yet it is useful.

I think this might belong to the profile. So a profile can have ``runtime = auto`` or ``runtime = MT auto`` or something like that.

File: conans/client/loader.py (hunk 1 above)
Comment (lasote) [inline conans/client/loader.py]: Why is the responsibility of the loader to mix the settings values with the settings?

`settings.values = profile.settings_values`

I agree with your observation but then let's extract it too. You need the values to preprocess the settings.

File: conans/client/manager.py (hunk 2 above)
Comment (lasote) [inline conans/client/manager.py]: You need the real settings, not the profile settings values. What if you want to remove the `os` in the preprocessing? https://github.com/conan-io/conan/pull/1565

File: conans/client/settings_preprocessor.py (hunk 4 above)
Comment (lasote) [inline conans/client/settings_preprocessor.py]: Then it has to be opt-out at least. We cannot force the user to specify MD => Release, MDd => Debug.
I would say that we shouldn't make that check.

File: conans/client/settings_preprocessor.py (hunk 3 above)
Comment (lasote) [inline conans/client/settings_preprocessor.py]: It's not the same to get "None" from an empty setting (no value in the profile) than: "That setting doesn't exist" (no setting in the setting.yml) and of course we cannot capture any exception not related with that.
We cannot just provide a generic and configurable setting.yml but then expect that the runtime (or the emscripten https://github.com/conan-io/conan/pull/1565) is declared.

The `auto` sounds good, but again, it's not only about the values, you could need to remove a setting in the preprocessing, it's not just a profile thing.

File: conans/client/loader.py (hunk 1 above)
Comment (memsharded) [inline conans/client/loader.py]: You can use the ``settings`` object, not just the values, to process it, augment it, while keeping its validation rules. Though it is true that it might break if someone is applying settings constraints. But you can apply the logic to the ``profile.settings``

**Comment by lasote** [comment]
I've pushed some changes with a refactor. Now the manager contains the settings preprocessor, the loader doesn't know anything about settings, values or preprocessing, only the package level values. Moved some functions to the manager and simplified things.

Please, take a look at my previous comments and let's keep discussing it.

File: conans/client/loader.py [hunk 5]
```diff
@@ -80,7 +31,7 @@ def __init__(self, runner, settings, profile):
         self._package_settings = profile.package_settings_values
         self._env_values = profile.env_values
 
-    def load_conan(self, conanfile_path, output, consumer=False, reference=None):
+    def load_conan(self, conanfile_path, scoped_output, consumer=False, reference=None):
```
Comment (memsharded) [inline conans/client/loader.py]: This rename to ``scoped_output`` is unnecessary

File: conans/client/settings_preprocessor.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+def fill_runtime(settings):
+    try:
+        if settings.compiler == "Visual Studio":
+            if settings.get_safe("compiler.runtime") is None:
+                settings.compiler.runtime = "MDd" if settings.get_safe("build_type") == "Debug" \
+                                                  else "MD"
+                logger.info("Setting 'compiler.runtime' not declared, automatically "
+                            "adjusted to '%s'" % settings.compiler.runtime)
+    except InvalidConanSettingField:  # If the settings structure doesn't match these general
```
Comment (memsharded) [inline conans/client/settings_preprocessor.py]: If the user has changed the settings.yml defaults, it can raise a "BadValue" exception, not a InvalidConanSettingField, and that would be breaking and blocking.

I think this "smart" setting should catch any exception to make sure, so InvalidConanSettingField might be unnecessary.

File: conans/test/integration/cmake_flags_test.py [hunk 7]
```diff
@@ -201,7 +201,8 @@ def build(self):
         client.save({"conanfile.py": conanfile % "True"})
         client.run("build", ignore_error=True)
 
-        self.assertIn("Execute 'conan install -g txt' first", client.user_io.out)
+        self.assertIn("self.deps_cpp_info not defined. If you need it for a local command run "
```
Comment (memsharded) [inline conans/test/integration/cmake_flags_test.py]: I am confused why this message has changed. I think the code about this is not related to this PR. I have difficulties comparing some refactors, movement of functions

File: conans/client/settings_preprocessor.py (hunk 6 above)
Comment (lasote) [inline conans/client/settings_preprocessor.py]: Ok, catch'em all then...

File: conans/test/integration/cmake_flags_test.py (hunk 7 above)
Comment (lasote) [inline conans/test/integration/cmake_flags_test.py]: I removed from the ``load_consumer_conanfile`` the:

```
if not profile:
    raise ConanException("Execute 'conan install -g txt' first.")
```

Because now it always receives the default profile (from the `self.client_cache.default_profile`). I unified the instantiation of the loader, there was a big mess of different instantiations with different parameters depending on the call. 
So I think it has the sense to get always the default profile because it always applies, right?

So, now, this test fails a little bit later, in `_load_deps_cpp_info`

Comment (memsharded) [inline conans/test/integration/cmake_flags_test.py]: Ok, perfect
